### PR TITLE
Add GET api to get shard routing weights

### PR DESCRIPTION
**Pull request by anshu1106**
### Description
 This pull request adds GET api to fetch weights for weighted routing policy.

### Issues Resolved
[List any issues this PR will resolve]
[2859](https://github.com/opensearch-project/OpenSearch/issues/2859)
[3639](https://github.com/opensearch-project/OpenSearch/issues/3639)
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/ClusterGetWRRWeightsResponse.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/ClusterGetWRRWeightsResponse.java]: Done

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java]: Removed

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java]: We can do that

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/ClusterGetWRRWeightsRequest.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/ClusterGetWRRWeightsRequest.java:50]: Added validations

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java]: Added validations. We would be returning proper exception in case attribute name is not proper.

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/wrr/get/TransportGetWRRWeightsAction.java]: We are sending empty response in case weights are not set.

**Comment by sachinpkale** [comment]
Overall changes look good. Will be able to approve only after PUT changes are merged and this PR is updated accordingly.

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java:117]: Please add unit tests for this new method

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:102]: if ```request.local()``` is true then we put local node weight in response in addition to attribute weights fetched from metadata. In all other cases, if weights are present in metadata then ```clusterGetWeightedRoutingResponse``` is initialized with weights else null values.

File: server/src/test/java/org/opensearch/cluster/action/shard/routing/weighted/get/ClusterGetWeightedRoutingResponseTests.java
Comment (sachinpkale) [inline server/src/test/java/org/opensearch/cluster/action/shard/routing/weighted/get/ClusterGetWeightedRoutingResponseTests.java:19]: Add tests for `ClusterGetWeightedRoutingRequest` as well.

File: server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java:117]: addressed

File: server/src/test/java/org/opensearch/cluster/action/shard/routing/weighted/get/ClusterGetWeightedRoutingResponseTests.java
Comment (anshu1106) [reply server/src/test/java/org/opensearch/cluster/action/shard/routing/weighted/get/ClusterGetWeightedRoutingResponseTests.java:19]: addressed

**Review by Bukhtawar — COMMENTED** [review]
Can we add integ tests for the API

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java]: Done

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java]: Addressed

**Comment by anshu1106** [comment]
> Can we add integ tests for the API

Added Integ tests

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:110]: explicit return is avoidable here

Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:110]: Sorry my bad. Fixing this

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:108]: we are throwing IOException what happens if we run into IOException?

**Review by Bukhtawar — APPROVED** [review]

File: server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java
Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:108]: returns 500 response in case of IOException

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java:108]: I meant will the listener be invoked or the request will be deemed stuck

Comment (anshu1106) [reply server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java]: Yeah correct, request can get stuck on this. Made change to invoke listener on all failures

**Comment by anshu1106** [comment]
Build failing due to unrelated test failures 
```[org.opensearch.client.PitIT.testDeleteAllAndListAllPits](https://build.ci.opensearch.org/job/gradle-check/3898/testReport/junit/org.opensearch.client/PitIT/testDeleteAllAndListAllPits/)
[org.opensearch.qa.verify_version_constants.VerifyVersionConstantsIT.testLuceneVersionConstant](https://build.ci.opensearch.org/job/gradle-check/3898/testReport/junit/org.opensearch.qa.verify_version_constants/VerifyVersionConstantsIT/testLuceneVersionConstant/)
[org.opensearch.qa.verify_version_constants.VerifyVersionConstantsIT.testLuceneVersionConstant](https://build.ci.opensearch.org/job/gradle-check/3898/testReport/junit/org.opensearch.qa.verify_version_constants/VerifyVersionConstantsIT/testLuceneVersionConstant_2/)
[org.opensearch.qa.verify_version_constants.VerifyVersionConstantsIT.testLuceneVersionConstant](https://build.ci.opensearch.org/job/gradle-check/3898/testReport/junit/org.opensearch.qa.verify_version_constants/VerifyVersionConstantsIT/testLuceneVersionConstant_3/)
[org.opensearch.qa.verify_version_constants.VerifyVersionConstantsIT.testLuceneVersionConstant](https://build.ci.opensearch.org/job/gradle-check/3898/testReport/junit/org.opensearch.qa.verify_version_constants/VerifyVersionConstantsIT/testLuceneVersionConstant_4/)```

Rebasing from main to see if that fixes the build

**Pull request merged**
